src/lib/statetable.rs

```diff
@@ -0,0 +1,158 @@
+use std::collections::hash_map::{Entry, HashMap};
+
+use grammar::{AIdx, Grammar, Symbol};
+use stategraph::StateGraph;
+
+/// A representation of a StateTable for a grammar. `actions` and `gotos` are split into two
+/// separate hashmaps, rather than a single table, due to the different types of their values.
+#[derive(Debug)]
+pub struct StateTable {
+    pub actions: HashMap<(usize, Symbol), Action>,
+    pub gotos  : HashMap<(usize, AIdx), usize>
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Action {
+    /// Shift to state X in the statetable.
+    Shift(usize),
+    /// Reduce alternative X in the grammar.
+    Reduce(usize),
+    /// Accept this input.
+    Accept
+}
+
+impl StateTable {
+    pub fn new(grm: &Grammar, sg: &StateGraph) -> StateTable {
+        let mut actions: HashMap<(usize, Symbol), Action> = HashMap::new();
+        let mut gotos  : HashMap<(usize, AIdx), usize>    = HashMap::new();
+
+        for (state_i, state) in sg.states.iter().enumerate() {
+            // Populate actions
+            for (&(alt_i, dot), ctx) in state.items.iter() {
+                let alt = &grm.alts[alt_i];
+                if dot < alt.len() {
+                    let sym = alt[dot].clone();
+                    if let Symbol::Terminal(_) = sym {
+                        if let Some(state_j) = sg.edges.get(&(state_i, sym)) {
+                            match actions.entry((state_i, sym)) {
+                                Entry::Occupied(e) => {
+                                    match e.get() {
+                                        &Action::Shift(x) => assert_eq!(*state_j, x),
+                                        &Action::Reduce(_) => panic!("shift/reduce error"),
```

Comment:
Ah, sorry, the PR should have said "we'll implement conflict resolution / detection later." I agree we definitely need it in the longer term, but I think that would be a good later PR.